architecture/authentication.adoc

```diff
@@ -0,0 +1,268 @@
+
+* `login`
+** Redirect to an OpenShift-provided login page, and validate the login using the configured link:#config-password[username and password validation]
+** For OAuth clients that request it, issue a `WWW-Authenticate: Basic` challenge rather than redirecting
+** Requires `session` and `basicauth` to be in the `ORIGIN_OAUTH_REQUEST_HANDLERS` list
+* `github`
+** Allows logging in using GitHub credentials
+** Uses the specified link:#config-github[Github OAuth client configuration]
+** Requires `session` to be in the `ORIGIN_OAUTH_REQUEST_HANDLERS` list
+** Using an external identity provider via OAuth prevents non-browser clients from obtaining API tokens directly
+* `google`
+** Allows logging in using Google credentials
+** Uses the specified link:#config-google[Google OAuth client configuration]
+** Requires `session` to be in the `ORIGIN_OAUTH_REQUEST_HANDLERS` list
+** Using an external identity provider via OAuth prevents non-browser clients from obtaining API tokens directly
+* `deny`
+** Reject unauthenticated requests
+
+`ORIGIN_OAUTH_HANDLER` defaults to `login`
+
+==== Bearer token validation [[config-bearer]]
+
+This configuration is used by the link:#config-request-handlers[`bearer`] request handler.
+
+To configure how the OAuth server validates bearer tokens, set the `ORIGIN_OAUTH_TOKEN_STORE` environment variable. Valid values are:
+
+* `etcd`
+** Bearer tokens must be valid OAuth access tokens (which are stored in etcd)
+** This allows a user with an API token to generate another API token
+* `file`
+** Bearer tokens are validated against a CSV file containing token, user name, scope, and user uid columns (scope and user uid may be blank). A sample row might look like this: `token123,bob,session,uid-bob`
+** Requires the `ORIGIN_OAUTH_TOKEN_FILE_PATH` environment variable be set to the path of the CSV file
+** The token file is read once at server start, not per request
+
+`ORIGIN_OAUTH_TOKEN_STORE` defaults to `etcd`
+
+==== Username and password validation [[config-password]]
+
+This configuration is used by the link:#config-request-handlers[`basicauth`] request handler and the link:#config-auth-handler[`login`] auth handler.
+
+To configure how usernames and passwords are validated, set the `ORIGIN_OAUTH_PASSWORD_AUTH` environment variable. Valid values are:
+
+* `anypassword`
+** Any non-empty username and password combination is accepted
+* `basicauthurl`
+** The username and password are validated against a protected JSON-returning remote URL, using basic auth:
+*** A 401 response indicates failed auth.
+*** A non-200 status or the presence of an "error" key with a non-empty value indicates an error: `{"error":"Error message"}`
+*** A 200 status with an "id" key indicates success: `{"id":"userid"}`
+*** A successful response may also include name and/or email: `{"id":"userid", "name": "User Name", "email":"user@example.com"}`
```

Comment:
Don't want to add them yet until we figure out how we add/persist groups from identity providers